keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,343 @@
+
+## Motivation
+
+Kubelet is responsible for running Health Checks (probes) and communicating the results via Pod status. All that information is stored in cache and reported to Kube-API. Right now pod information is tightly coupled with the Kubernetes API Server. When a workload wants to know the actual state of pods running on the Node, it needs to fetch it from Kube-API.
+This causes some issues:
+Reliability - for various reasons Kube-API might not be available (connectivity issue, control plane updates)
+Scalability - adding new watchers to kube-API is a scalability concern
+By exposing the endpoint that will serve the pods status directly from the kubelet cache we can use  it on node workloads and avoid additional dependencies on the Kubernetes API Server.
+
+| Impact | Description|
+| ------- | ------------ |
+| + | Reliability - for various reasons kube-API might not be available but this doesn’t mean that local workloads are not accessible and on node system workloads should have the most recent data about pod's readiness even when kube-API is unreachable. |
+| + | Scalability - Reduce the load on kube-API be reducing the number of watchers and using kubelet to fetch local pods data. Fetching only pods limited to one node is costly operation for kube-API. |
+| + | Safety - Read-only API will not add security risks. |
+| - | This API will add load to kubelet (mitigation: API will be rate limited) |
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is design to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the source - kubelet. This would remove the need for node-local components to request this node-local information from the Kubernetes API Server. Kubelet already has a podresources endpoint ([2403-pod-resources-allocatable-resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources)) which returns information about Pod’s containers and Devices. This API does not contain information about pod readiness status. Kubelet is responsible for computing pod status and stores it in a local cache. We want to create a new API that will expose pod conditions that are computed by kubelet and return the most recent data even when kube-API is not reachable.
+
+### Non-Goals
+
+### User Stories
+* Some on node system workloads want to reduce Control Plane dependency and introduce locality for pod’s readiness due to reliability and scalability issues.
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
```

Comment:
I for myself I'll need to review the past conversations about the topic which KEP wants to cover. This is my biggest concern. If exposing this data is ok at architecture (sig-node/sig-arch) level, everything else I see here seems pretty sensible straightforward to me, pending more review cycles.
I'm not sure about the consequences of tolerating control plane unavailabvility and effectively duplicate this information.

Reply:
Yep, we've been discussing this proposal at both sig-node and sig-arch meetings (this is on the agenda for tomorrow's sig-arch meeting if you'd like to join).

From my biased perspective, I think it's important to remove unnecessary dependencies on the control plane wherever we can. In the case of health checking in particular, this information is originating from Kubelet, so a node-local component fetching this from Kubelet would have 2 advantages:

* Reduced load on control plane
* Improved reliability if/when control plane is unavailable